src/exec/command.rs

```diff
@@ -6,47 +6,93 @@ use std::sync::Mutex;
 use crate::error::print_error;
 use crate::exit_codes::ExitCode;
 
+struct Outputs {
+    stdout: Vec<u8>,
+    stderr: Vec<u8>,
+}
+struct OutputBuf<'a> {
+    out_perm: &'a Mutex<()>,
+    outputs: Vec<Outputs>,
+}
+
+impl<'a> OutputBuf<'a> {
+    fn new(out_perm: &'a Mutex<()>) -> Self {
+        Self {
+            out_perm,
+            outputs: Vec::new(),
+        }
+    }
+
+    fn push(&mut self, stdout: Vec<u8>, stderr: Vec<u8>) {
+        self.outputs.push(Outputs { stdout, stderr });
+    }
+
+    fn write(self) {
+        // avoid taking the lock if there is nothing to do
+        if self.outputs.is_empty() {
+            return;
+        }
+        // While this lock is active, this thread will be the only thread allowed
+        // to write its outputs.
+        let _lock = self.out_perm.lock().unwrap();
+
+        let stdout = io::stdout();
+        let stderr = io::stderr();
+
+        let mut stdout = stdout.lock();
+        let mut stderr = stderr.lock();
+
+        for output in self.outputs.iter() {
+            let _ = stdout.write_all(&output.stdout);
+            let _ = stderr.write_all(&output.stderr);
+        }
+    }
+}
+
 /// Executes a command.
-pub fn execute_command(
-    mut cmd: Command,
+pub fn execute_commands<I: Iterator<Item = Command>>(
+    cmds: I,
     out_perm: &Mutex<()>,
     enable_output_buffering: bool,
 ) -> ExitCode {
-    // Spawn the supplied command.
-    let output = if enable_output_buffering {
-        cmd.output()
-    } else {
-        // If running on only one thread, don't buffer output
-        // Allows for viewing and interacting with intermediate command output
-        cmd.spawn().and_then(|c| c.wait_with_output())
-    };
-
-    // Then wait for the command to exit, if it was spawned.
-    match output {
-        Ok(output) => {
-            // While this lock is active, this thread will be the only thread allowed
-            // to write its outputs.
-            let _lock = out_perm.lock().unwrap();
-
-            let stdout = io::stdout();
-            let stderr = io::stderr();
-
-            let _ = stdout.lock().write_all(&output.stdout);
-            let _ = stderr.lock().write_all(&output.stderr);
-
-            if output.status.code() == Some(0) {
-                ExitCode::Success
-            } else {
-                ExitCode::GeneralError
+    let mut out_buf = OutputBuf::new(out_perm);
```

Comment:
I would appreciate if we could write out all terms here:

* out_buf => output_buffer
* OutputBuf => OutputBuffer
* out_per => output_permission (I realize this was there already)